Ruby_Tutorial/bst.rb

```diff
@@ -0,0 +1,228 @@
+		puts current.value.to_s
+	end
+
+	def inorder(current = @root)
+		return if (current == nil)
+		inorder(current.left)
+		puts current.value.to_s
+		inorder(current.right)
+	end
+
+	def level_order(current = @root)
+		return if (current == nil)
+		puts current.value.to_s
+		level_order(current.left)
+		level_order(current.right)
+	end
+
+	def data
+		@data ||= []
+	end
+
+	def store_elements(current = @root)
+		return if (current == nil)
+		data << current.value.to_i
+		store_elements(current.left)
+		store_elements(current.right)
+	end
+
+	def delete(current = @root, value, parent)
+		if (root ==  nil)
+			return root
+		elsif (value < current.value)
+			delete(current.left, value, current)
+		elsif (value > current.value)
+			delete(current.right, value, current)
+		else
+		#case 1 if node has no child
+			if current.left == nil && current.right == nil
+				if value == parent.value
+					current = nil
+				elsif (value < parent.value)
+					parent.left = nil
+				else
+					parent.right = nil
+				end
+		#case 2 if the node has a single child
+			elsif current.left == nil
+				if (value < parent.value)
+					parent.left = current.right
+				else
+					parent.right = current.right
+				end
+			elsif current.right == nil
+				if (value < parent.value)
+					parent.left = current.left
+				else
+					parent.right = current.left
+				end
+		#case 3 if node has 2 childs
+			else
+				@minval = self.smallest(current)
+				temp = @minval.value
+				delete(temp, nil)
+				current.value = temp
+			end
+		end
+	end
+
+	def root_to_leaf(current = @root, path = [], path_collection = [])
+		return if current.nil?
+		path << current.value
+		if current.left == nil && current.right == nil
+			path_collection << path.dup
+			print path
+			puts "path is"
+			return
+		end
+		root_to_leaf(current.left, path, path_collection)
+		path.pop
+		root_to_leaf(current.right, path , path_collection)
+	end
+
+end
+
+class Node
+	attr_accessor :left, :right, :value
+
+	def initialize(left = nil, right = nil, value)
+		@left = left
+		@right = right
+		@value = value
+	end
+end
+
+bst = BinarySearchTree.new(10)
+
+puts "Enter your choice -- Type help to see all commands"
+while (choice = gets.chomp)
+	if choice == 'help'
+		puts "Commands          Description"
+		puts "insert            -- to add elements to bst"
+		puts "largest           -- to print the largest element"
+		puts "smallest          -- to print the smallest element"
+		puts "inorder           -- to print the Inorder"
+		puts "postorder         -- to print the Postorder"
+		puts "level_order       -- to print the Levelorder"
+		puts "preorder          -- to print the preorder"
+		puts "search            -- to search if a element is present"
+		puts "remove            -- to remove an element"
+		puts "path              -- to print all path from root to leaf"
+		puts "quit              -- to quit, all elements are stored in a file"
+		puts "load              -- to read input file and insert element in bst"
```

Comment:
create a module of constants, to avoid string comparison without losing readability.